bpf-gpl/nat.h

```diff
@@ -154,18 +154,29 @@ static CALI_BPF_INLINE struct calico_nat_dest* calico_v4_nat_lookup2(__be32 ip_s
 			return NULL;
 		}
 
-		/* XXX replace the following with a nodeport cidrs lookup once
-		 * XXX we have it.
-		 */
-		rt = cali_rt_lookup(ip_dst);
-		if (!rt) {
-			CALI_DEBUG("NAT: route miss\n");
-			return NULL;
-		}
-
-		if (!cali_rt_is_host(rt)) {
-			CALI_DEBUG("NAT: route dest not a host\n");
-			return NULL;
+		if (!from_tun) {
+			/* XXX replace the following with a nodeport cidrs lookup once
+			 * XXX we have it.
+			 */
+			rt = cali_rt_lookup(ip_dst);
+			if (!rt) {
+				CALI_DEBUG("NAT: route miss\n");
+				return NULL;
+			}
+
+			if (!cali_rt_is_host(rt)) {
+				CALI_DEBUG("NAT: route dest not a host\n");
+				return NULL;
+			}
+		} else {
+			/* we got here because the original node that forwarded
+			 * it through the tunnel though it is a nodeport, we can
+			 * use the wildcard nodeport entry.
+			 *
+			 * If the nodes have multiple IPs/NICs, RT entries would
+			 * not know the other IPs of other nodes.
+			 */
+			CALI_DEBUG("NAT: skip route lookup from tunnel, it is a nodeport\n");
```

Comment:
I think this misses a case:

- external node connects to new service cluster IP via node 1
- node 1 tunnels to node 2
- node 2 hasn't heard about cluster IP yet, gets NAT miss, treats it as nodeport when it shouldn't

Reply:
That is true. Unless we know that the other node has 2 IPs we either cannot handle that case or we need to live with the possible incorrectness.

I can rephrase the now relaxed condition to do exactly what it used to do plus if there is an RT miss because in the multi NIC scenario because we do not know that the other host has also another IP, we can then treat it as nodeport. This would still make a possible false assumption about a nodeport.

Reply:
I tried to adjust the condition, but not sure if it is any better. It should be subject of a runtime configuration - eventually.